File: Classes/GameEngine/Entities/Button.m
//
//  Button.m
//  AirHockey
//
//  Created by Jonathan Sharkey on 10-04-30.
//  Copyright 2010 Sharkable. All rights reserved.
//

#import "Button.h"
#import "SoundPlayer.h"
#import "ResourceLoader.h"

@implementation Button

@synthesize delegate=delegate_, selector=selector_;

- (id)initWithNormalTexture:(Texture2D *)normalTexture
             pressedTexture:(Texture2D *)pressedTexture
                   position:(CGPoint)position {
  [super init];
  
  if (self) {
    normalTexture_ = normalTexture;  
    pressedTexture_ = pressedTexture;
    
    position_ = position;
    state_ = BUTTON_STATE_NORMAL;
  }
  
  return self;
}

- (void)dealloc {
  [[ResourceLoader instance] releaseResource:normalTexture_];
  [[ResourceLoader instance] releaseResource:pressedTexture_];
  
  [super dealloc];
}

- (void) update {
}

- (void)render {
  switch (state_) {
    case BUTTON_STATE_NORMAL: {
      [normalTexture_ drawAtPoint:position_];
      break;
    }
    case BUTTON_STATE_PRESSED: {
      [pressedTexture_ drawAtPoint:position_];
      break;
    }
  }
}

- (void)touchesBegan:(Touch *[])touches numTouches:(int)numTouches {
  if (state_ == BUTTON_STATE_NORMAL) {
    for (int i = 0; i < numTouches; i++) {
      if ([self containsPoint:touches[i].location]) {
        state_ = BUTTON_STATE_PRESSED;
        [SoundPlayer playSound:kSoundButton];
      }
    }
  }
}

- (void)touchesEnded:(Touch *[])touches numTouches:(int)numTouches {
  if (state_ == BUTTON_STATE_PRESSED) {
    state_ = BUTTON_STATE_NORMAL;
    for (int i = 0; i < numTouches; i++) {
      if ([self containsPoint:touches[i].location]) {
        [delegate_ performSelector:selector_];
      }
    }
  }
}

- (BOOL)containsPoint:(CGPoint)p {
  return p.x >= position_.x &&
       p.y >= position_.y &&
       p.x < position_.x + pressedTexture_.contentSize.width &&
       p.y < position_.y + pressedTexture_.contentSize.height;
}

- (CGSize) size {
  return normalTexture_.contentSize;
}

@end

File: Classes/GameEngine/Entities/button.cc
//
//  button.m
//  AirHockey
//
//  Created by Jonathan Sharkey on 10-04-30.
//  Copyright 2010 Sharkable. All rights reserved.
//

#include "gameengine/entities/button.h"

#include "gameengine/resource_loader.h"
#include "gameengine/touch.h"
#include "soundengine/sound_player.h"

Button::Button(Texture2D normal_texture, Texture2D pressed_texture, GamePoint position)
    : normal_texture_(normal_texture),
      pressed_texture_(pressed_texture),
      position_(position),
      state_(kButtonStateNormal),
      delegate_(NULL) {
}

Button::~Button() {
  if (normal_texture_.loaded()) {
    ResourceLoader::Instance().ReleaseResource(normal_texture_);
  }
  if (pressed_texture_.loaded()) {
    ResourceLoader::Instance().ReleaseResource(pressed_texture_);
  }
}

bool Button::ContainsPoint(GamePoint p) {
  return p.x >= position_.x &&
  p.y >= position_.y &&
  p.x < position_.x + pressed_texture_.content_size().width &&
  p.y < position_.y + pressed_texture_.content_size().height;
}


// ViewEntity

void Button::Update() {
}

void Button::Render() {
  switch (state_) {
    case kButtonStateNormal: {
      // TODONOW normal_texture_.DrawAtPoint(position_);
      break;
    }
    case kButtonStatePressed: {
      // TODONOW pressed_texture_.DrawAtPoint(position_);
      break;
    }
  }
}

void Button::TouchesBegan(vector<Touch> touches) {
  if (state_ == kButtonStateNormal) {
    for (int i = 0; i < touches.size(); i++) {
// TODONOW
//      if (ContainsPoint(touches[i].location())) {
//        state_ = kButtonStatePressed;
//        SoundPlayer::instance()->playSound(kSoundButton);
//      }
    }
  }
}

void Button::TouchesEnded(vector<Touch> touches) {
  if (state_ == kButtonStatePressed) {
    state_ = kButtonStateNormal;
    for (int i = 0; i < touches.size(); i++) {
// TODONOW
//      if (ContainsPoint(touches[i].location())) {
//        if (delegate_) {
//          delegate_->ButtonPressed(this);
//        }
//      }
    }
  }
}
